xlators/mgmt/glusterd/src/glusterd.h

```diff
@@ -247,8 +247,9 @@ typedef struct glusterd_add_dict_args {
 } glusterd_add_dict_args_t;
 
 typedef struct glusterd_friend_synctask_args {
-    char *dict_buf;
-    u_int dictlen;
+    dict_t *peer_data;
+    dict_t *peer_ver_data;          // Dictionary to save peer version data
+    uint64_t status_arr[256];       // Array to save volume update status
```

Comment:
So, this now makes the assumption 16k (64x256) volumes are maximum with a single cluster. While that is a lot compared to where we are, we need to specifically call it out.

Reply:
> So, this now makes the assumption 16k (64x256) volumes are maximum with a single cluster. While that is a lot compared to where we are, we need to specifically call it out.

The function glusterd_add_volume_to_dict we do save almost 40 key-value per volume and in case of 4k volumes the total keys will be around 1.6L.As we know in current  dictionary we do save all the keys value pair as a linked list so at the time of  comparing key-value to fetch the values it takes time.IMO even if we use gfx_dict we can't save time to access the keys.With gfx_dict we can save time to serialize/unserialize the dictionary that is not much.
There are two main changes i did to optimize it
      1) To take a decision about peer volume has changed we need only 5 keys so i have 
      saved those 5 keys in a specific dictionary so for total 4k volumes we will have 20k keys.
      
   2) If peer has some updates instead of updating status in some dictionary i have introduced
      a new arr(status_arr) and based on the single bit access we can take decision a peer has
      some update and need to import it.
After applied both changes i am able to access 5k volume updates during handshake easily otherwise it is not possible.

Reply:
@amarts Can you please give your vote on this.